stl/inc/xutility

```diff
@@ -4124,12 +4124,51 @@ _CONSTEXPR20 _OutIt _Copy_unchecked(_InIt _First, _InIt _Last, _OutIt _Dest) {
     return _Dest;
 }
 
+// _Iterators_are_contiguous<_Iter1, _Iter2> reports whether both iterators are known to be contiguous.
+// (Without concepts, this detection is limited, which will limit when we can activate the memcmp optimization.)
+
+#ifdef __cpp_lib_concepts
+// When concepts are available, we can detect arbitrary contiguous iterators.
+template <class _Iter1, class _Iter2>
+_INLINE_VAR constexpr bool _Iterators_are_contiguous = contiguous_iterator<_Iter1> //
+    && contiguous_iterator<_Iter2>;
+
+template <class _Target, class _Iter>
+_NODISCARD constexpr _Target* _To_pointer(const _Iter& _It) noexcept {
+    _STL_INTERNAL_STATIC_ASSERT(contiguous_iterator<_Iter>);
+    return reinterpret_cast<_Target*>(_STD to_address(_It));
+}
+#else // ^^^ defined(__cpp_lib_concepts) ^^^ / vvv !defined(__cpp_lib_concepts) vvv
+// When concepts aren't available, we can detect pointers. (Iterators should be unwrapped before using this.)
+template <class _Iter1, class _Iter2>
+_INLINE_VAR constexpr bool _Iterators_are_contiguous = conjunction_v<is_pointer<_Iter1>, is_pointer<_Iter2>>;
+
+template <class _Target, class _Iter>
+_NODISCARD constexpr _Target* _To_pointer(const _Iter& _It) noexcept {
+    _STL_INTERNAL_STATIC_ASSERT(is_pointer_v<_Iter>);
+    return reinterpret_cast<_Target*>(_It);
+}
+#endif // ^^^ !defined(__cpp_lib_concepts) ^^^
+
 template <class _InIt, class _OutIt>
 _CONSTEXPR20 _OutIt copy(_InIt _First, _InIt _Last, _OutIt _Dest) { // copy [_First, _Last) to [_Dest, ...)
     _Adl_verify_range(_First, _Last);
     const auto _UFirst = _Get_unwrapped(_First);
     const auto _ULast  = _Get_unwrapped(_Last);
     const auto _UDest  = _Get_unwrapped_n(_Dest, _Idl_distance<_InIt>(_UFirst, _ULast));
+
+#if _ITERATOR_DEBUG_LEVEL == 2
+    if constexpr (_Iterators_are_contiguous<_Unwrapped_t<_InIt>, _Unwrapped_t<_OutIt>>) {
+#ifdef __cpp_lib_is_constant_evaluated
+        if (!_STD is_constant_evaluated())
+#endif // __cpp_lib_is_constant_evaluated
+        {
+            _STL_VERIFY(
+                _UDest + (_ULast - _UFirst) <= _UFirst || _ULast <= _UDest, "ranges should not overlap each other");

```

Comment:
This is guarded by `_Iterators_are_contiguous<_Unwrapped_t<_InIt>, _Unwrapped_t<_OutIt>>`, but the iterators might not be comparable. For example, a range of `short` can be copied to a range of `int`, yet `short*` can't be compared to `int*`.

Note that requiring the unwrapped iterator types to be the same would be too strict - a range of `const int` can be copied to a range of `int`, and `const int*` can be compared to `int*`.

Additionally, this indicates that we're missing test coverage for different-element-types scenarios.

Reply:
Sorry I missed these scenarios. I fixed it using type-traits and existing helper functions. I hope this would be correct.
Also, I've added some a test to copying different-element-types scenarios. I hope I have chosen the right place for these tests.